packages/beacon-node/src/api/impl/validator/index.ts

```diff
@@ -651,47 +794,60 @@ export function getValidatorApi(
 
     // Start calls for building execution and builder blocks
 
-    const builderPromise = isBuilderEnabled
-      ? produceBuilderBlindedBlock(slot, randaoReveal, graffiti, {
-          feeRecipient,
-          // can't do fee recipient checks as builder bid doesn't return feeRecipient as of now
-          strictFeeRecipientCheck: false,
-          // skip checking and recomputing head in these individual produce calls
-          skipHeadChecksAndUpdate: true,
-          commonBlockBody,
-          parentBlockRoot,
-        })
+    const builderBodyPromise = isBuilderEnabled
+      ? produceBuilderBlockBody({...blockAttributes, currentState})
       : Promise.reject(new Error("Builder disabled"));
 
-    const enginePromise = isEngineEnabled
-      ? produceEngineFullBlockOrContents(slot, randaoReveal, graffiti, {
-          feeRecipient,
-          strictFeeRecipientCheck,
-          // skip checking and recomputing head in these individual produce calls
-          skipHeadChecksAndUpdate: true,
-          commonBlockBody,
-          parentBlockRoot,
-        }).then((engineBlock) => {
+    const engineBodyPromise = isEngineEnabled
+      ? produceEngineBlockBody({...blockAttributes, currentState}).then((engineBlockBody) => {
           // Once the engine returns a block, in the event of either:
           // - suspected builder censorship
           // - builder boost factor set to 0 or builder selection `executionalways`
           // we don't need to wait for builder block as engine block will always be selected
           if (
-            engineBlock.shouldOverrideBuilder ||
+            engineBlockBody.shouldOverrideBuilder ||
             builderBoostFactor === BigInt(0) ||
             builderSelection === routes.validator.BuilderSelection.ExecutionAlways
           ) {
             controller.abort();
           }
-          return engineBlock;
+          return engineBlockBody;
         })
       : Promise.reject(new Error("Engine disabled"));
 
-    const [builder, engine] = await resolveOrRacePromises([builderPromise, enginePromise], {
-      resolveTimeoutMs: cutoffMs,
-      raceTimeoutMs: BLOCK_PRODUCTION_RACE_TIMEOUT_MS,
-      signal: controller.signal,
-    });
+    const builderTimer = metrics?.blockProductionTime.startTimer();
+    const engineTimer = metrics?.blockProductionTime.startTimer();
+
+    const [_, [builder, engine]] = await Promise.all([
+      commonBlockBodyPromise,
+      resolveOrRacePromises([builderBodyPromise, engineBodyPromise], {
+        resolveTimeoutMs: cutoffMs,
+        raceTimeoutMs: BLOCK_PRODUCTION_RACE_TIMEOUT_MS,
+        signal: controller.signal,
+      }),
+    ])
+      .then(async ([commonBlockBody, [builderBlockBody, engineBlockBody]]) => {
```

Comment:
it's more straightforward to get this out of `then` and move the whole code here to the 1st level